module/zfs/zio_compress.c

```diff
@@ -115,11 +115,28 @@ zio_compress_data(enum zio_compress c, abd_t *src, void *dst, size_t s_len)
 	if (c == ZIO_COMPRESS_EMPTY)
 		return (s_len);
 
+	/* No compression algorithms can read from ABDs directly */
+	void *tmp = abd_borrow_buf_copy(src, s_len);
+
+	/* for gzip 6+, use lz4 as a heuristic for compressibility */
+	if (ci->ci_compress == gzip_compress && ci->ci_level >= 6) {
+	    /* lz4 must find some compression */
+            if (s_len) {
+                d_len = s_len - 1;
+            } else {
+                d_len = s_len;
+            }
```

Comment:
For this to be worthwhile we need to save at least a full sector since that's what it will be padded out too.  Let's go ahead and assume a 4k sector size since the actual ashift isn't available, and this is only a heuristic.  So something like this which lets you drop the `if` conditional.

```c
d_len = MAX(s_len - 4096, s_len);
```

Reply:
Making the assumption of 4K sectors would mean that gzip is mostly disabled on small-block files/zvols, right?  That seems like an unnecessary regression in functionality.  Could we somehow have the spa_t accessible here, so that we could check the spa_min_ashift?

Reply:
That would be good, and after a quick glance it does look like it's available in all the callers so it can be passed in.

Reply:
@behlendorf this need also an additional check, if the data-size before compression is actually smaller than one sector size, since it might compress with gzip that well, that it can be stored in the block pointer by the embedded_data, right?

Reply:
@RubenKelevra If the data is all zeros, it will be caught by the zio_compress_zeroed_cb() check on line 112, and written as a hole rather than as data.

The point @behlendorf was trying to make is that the SAVINGS needs to be at least one sector, else doing any compression is a waste, since it won't actually consume any less space on disk.

Reply:
@RubenKelevra it's probably best to leave review to developers at this point.

Reply:
@kpande I don't feel comfortable with your replies anymore. Just stop.

Reply:
@allanjude well, my point was: if the data is just one sector, no sector can be saved anymore. So the check will always return zero sectors saved. If the data can be compressed very well, but is also very small uncompressed, it won't be compressed anymore and thus the embedded_data feature isn't triggered.

Reply:
@kpande: I think that @RubenKelevra is engaging in a technical discussion about the code in question.  One needn't be an expert to ask questions and try to understand the code.